src/cargo/core/resolver/dep_cache.rs

```diff
@@ -167,34 +167,63 @@ impl<'a> RegistryQueryer<'a> {
     }
 }
 
-pub fn build_deps(
-    registry: &mut RegistryQueryer<'_>,
-    parent: Option<PackageId>,
-    candidate: &Summary,
-    method: &Method,
-) -> ActivateResult<(Vec<String>, HashSet<InternedString>, Vec<DepInfo>)> {
-    // First, figure out our set of dependencies based on the requested set
-    // of features. This also calculates what features we're going to enable
-    // for our own dependencies.
-    let (warnings, used_features, deps) = resolve_features(parent, candidate, method)?;
-
-    // Next, transform all dependencies into a list of possible candidates
-    // which can satisfy that dependency.
-    let mut deps = deps
-        .into_iter()
-        .map(|(dep, features)| {
-            let candidates = registry.query(&dep)?;
-            Ok((dep, candidates, Rc::new(features)))
-        })
-        .collect::<CargoResult<Vec<DepInfo>>>()?;
-
-    // Attempt to resolve dependencies with fewer candidates before trying
-    // dependencies with more candidates. This way if the dependency with
-    // only one candidate can't be resolved we don't have to do a bunch of
-    // work before we figure that out.
-    deps.sort_by_key(|&(_, ref a, _)| a.len());
-
-    Ok((warnings, used_features, deps))
+pub struct DepsCache<'a> {
+    pub registry: RegistryQueryer<'a>,
+    cache: HashMap<
+        (Option<PackageId>, Summary, Method),
+        Rc<(Vec<String>, HashSet<InternedString>, Rc<Vec<DepInfo>>)>,
+    >,
+}
+
+impl<'a> DepsCache<'a> {
+    pub fn new(registry: RegistryQueryer<'a>) -> Self {
+        DepsCache {
+            registry,
+            cache: HashMap::new(),
+        }
+    }
+
+    pub fn build_deps(
+        &mut self,
+        parent: Option<PackageId>,
+        candidate: &Summary,
+        method: &Method,
+    ) -> ActivateResult<Rc<(Vec<String>, HashSet<InternedString>, Rc<Vec<DepInfo>>)>> {
+        if let Some(out) = self
```

Comment:
Could a comment be added here as to why we can cache the results? It'd be good to explain why this cache works the way it does and what enables it (e.g. the things that don't change and how this is a "pure" query which can be memoized.